### PR TITLE
Add untrash order hook (COT/HPOS).

### LINKED ISSUE #35018: [COT/HPOS]: Add corresponding hooks for `untrash_post` for COT/HPOS implementation
### Describe the solution you'd like

I am working on the Deposits extension to add support for trash, untrash, and delete orders for COT/HPOS implementation. To handle, when a parent order is trashed related scheduled orders should get trashed, when parent order is untrashed/restored, related scheduled orders should get untrashed/restored and finally when parent order deleted, related parent order should get deleted as well.

We have added corresponding hooks for `wp_trash_post` and `delete_post` for COT implementation in PR https://github.com/woocommerce/woocommerce/pull/34858, but we still don't have support for `untrash_post` which I needed for restored related scheduled payment orders when parent order restored/untrashed.

### Describe alternatives you've considered

Alternatively, we may use the `woocommerce_order_status_changed` hook to check if the order gets updated from `trash` status to something else. 

### Additional context

Deposits issue https://github.com/woocommerce/woocommerce-deposits/issues/467

### PR DESCRIPTION
**Pull request by barryhughes**
Adds a new action hook: `woocommerce_untrash_order`.

- This is an analog of the core WP hook `untrash_post`, but applies to (COT/HPOS) orders.
- Naming convention follows the hooks we added in https://github.com/woocommerce/woocommerce/pull/34858.
- Essentially replicates the same params as provided by `untrash_post`.

Closes #35018.

### How to test the changes in this Pull Request:

Reviewing the code should be sufficient (by itself, this does not introduce any functional changes), but you could also add a snippet like this to a suitable location:

```php
add_action( 'woocommerce_untrash_order', function ( int $order_id ) {
	wc_get_logger()->info( "woocommerce_untrash_order fired for order #{$order_id}." );
} );
```

Then, after trashing and untrashing an order, look in the logs found in **WooCommerce ▸ Status ▸ Logs** and you should see an entry referencing the order you just untrashed.

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> run changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by barryhughes** [comment]
Keeping as a draft, there are some other problems with trashing and deleting (when COT is authoritative and sync is enabled) that are getting in the way of this change.

**Comment by barryhughes** [comment]
Some work is in progress re the other issues I noted; figured there was little value in holding this work back since it simply adds a hook (though, you won't be able to test as per the PR description until those other problems are solved).

**Comment by jeffpaul** [comment]
Tagging @peterfabian @vedanshujain to help get this PR finalized/merged to unblock work on Bookings and Deposits compatibility work for HPOS... please and thanks!

**Review by jorgeatorres — APPROVED** [review]
LGTM. Thank you @barryhughes!

FWIW, I tested by rebasing this branch off of `fix/35089` first (so that trashing/untrashing works) and then adding the custom code to confirm the hook is fired at the correct time.

**Pull request merged**
